### PR TITLE
Dex_xcmp: get rid of generic_asset, new downward/xcmp message handlers draft implementation

### PR DESCRIPTION
**Pull request by iorveth**
The scope of this PR includes:

1. Downward/xcmp message handlers implementation, working  both with native currency and custom token assets transfer messages.
1. Extrinsics for transferring  both natively supported currency and custom tokens from `dex_parachain`.

### CODE REVIEW

File: pallets/dex-xcmp/src/lib.rs [hunk 1]
```diff
@@ -117,139 +137,210 @@ decl_module! {
...
             XCMPMessage::TransferToken(dest, amount) => {
-                let dex_account_id = Self::dex_account_id();
-
-                if !<CurrencyByParaId<T>>::contains_key(&src) {
-                    let asset_id = <generic_asset::Module<T>>::next_asset_id();
-
-                    let dex_asset_options = Self::create_default_dex_asset_options();
+                <dex_pallet::Module<T>>::ensure_can_hold_balance(
+                    &dest,
+                    <T as dex_pallet::Trait>::KSMAssetId::get(),
```
Comment (whalelephant) [inline pallets/dex-xcmp/src/lib.rs]: The purpose of `XCMPMessage::TransferToken` was for parachains without multiassets. 
But if receive on dex-chain we mint `KSMAssetId` tokens, then someone can send this to Dex-chain and withdraw KSM directly, but our dex account on relay chain does not have those KSM. 

So maybe like what you had -`ensure_asset_id_exist(src, some_asset_id_)`, where `some_asset_id` is considered as main_currency, at least to dex-chain, so we at least some can keep track of it. Or maybe we do not handle this type of message at all.

File: pallets/dex-xcmp/src/lib.rs [hunk 2]
```diff
@@ -117,139 +137,210 @@ decl_module! {
...
+        //
+        // == MUTATION SAFE ==
+        //
+
+        match msg {
+            XCMPMessage::TransferToken(dest, amount) => {
+                <dex_pallet::Module<T>>::mint_asset(
                     &dest,
-                    &amount,
+                    <T as dex_pallet::Trait>::KSMAssetId::get(),
```
Comment (whalelephant) [inline pallets/dex-xcmp/src/lib.rs]: same as above

File: pallets/dex-xcmp/src/lib.rs [hunk 3]
```diff
@@ -117,139 +137,210 @@ decl_module! {
...
+            <dex_pallet::Module<T>>::ensure_sufficient_balance(&who, <T as dex_pallet::Trait>::KSMAssetId::get(), amount)?;
 
-            <generic_asset::Module<T>>::burn_free(
-                &asset_id, &Self::dex_account_id(), &who, &amount
-            )?;
+            //
+            // == MUTATION SAFE ==
+            //
+
+            <dex_pallet::Module<T>>::slash_asset(&who, <T as dex_pallet::Trait>::KSMAssetId::get(), amount);
```
Comment (whalelephant) [inline pallets/dex-xcmp/src/lib.rs]: please see comment below regarding remote parachain main currency handling

**Review by whalelephant — CHANGES_REQUESTED** [review]

File: pallets/dex-xcmp/src/lib.rs (hunk 1 above)
Comment (iorveth) [reply pallets/dex-xcmp/src/lib.rs]: Good point! The issue here is that i can't be sure regarding the asset_ids on other parachains to say exactly which one corresponds to main network currency, like `ksm`. 
I want to have some way to distinguish between natively supported currencies, main assets on other parachains and custom assets, that we keep track remotely. 
Currently we have a native support only for one currency, let's say `KSM` for  kusama/ `DOT` for polkadot network respectively. 
We can introduce `transferKsm` message for that case and refactor `TransferToken` to reflect the semantic of main currency from other parachain the way you suggest.

**Pull request merged**